sandbox/lua/encoders/opentsdb_http.lua

```diff
@@ -0,0 +1,190 @@
+      ts_from_message = true
+      add_hostname_if_missing = true
+      type_as_prefix = true
+      tag_prefix = "tag_"
+      tag_list = "env dc"
+      skip_fields = "Pid"
+
+    [opentsdb]
+    type = "HttpOutput"
+    message_matcher = "Fields[MetricOpentsdb] != NIL"
+    address = "http://opentsdb.example.com:4242/api/put"
+    encoder = "OpentsdbEncoder"
+
+*Example Output*
+.. code-block:: json
+    [{"value":0.05,"timestamp":1443535912,"metric":"stats.loadavg.15MinAvg","tags":{"env":"test","host":"example.com"}},{"value":2,"timestamp":1443535912,"metric":"stats.loadavg.NumProcesses","tags":{"env":"test","host":"example.com"}},{"value":0.01,"timestamp":1443535912,"metric":"stats.loadavg.5MinAvg","tags":{"env":"test","host":"example.com"}},{"value":0,"timestamp":1443535912,"metric":"stats.loadavg.1MinAvg","tags":{"env":"test","host":"example.com"}}]
+
+--]=]
+
+require "cjson"
+require "table"
+require "string"
+require "os"
+
+local l = require 'lpeg'
+l.locale(l)
+
+local flush_count     = read_config("flush_count") or 1
+local flush_delta     = read_config("flush_delta") or 0
+local tag_prefix      = read_config("tag_prefix")
+local tag_list_str    = read_config("tag_list") or ""
+local skip_fields_str = read_config("skip_fields") or ""
+local type_as_prefix  = read_config("type_as_prefix")
+local ts_from_message = read_config("ts_from_message")
+local add_hostname    = read_config("add_hostname_if_missing")
+_PRESERVATION_VERSION = read_config("preservation_version") or 1
+
+if not tag_prefix then
+  return -1
+end
+
+if type_as_prefix == nil then
+  type_as_prefix = true
+end
+
+if add_hostname == nil then
+  add_hostname = true
+end
+
+if ts_from_message == nil then
+  ts_from_message = true
+end
+
+tag_list = {}
+for field in tag_list_str:gmatch("[%S]+") do
+  tag_list[field] = true
+end
+
+skip_fields = {}
+for field in skip_fields_str:gmatch("[%S]+") do
+  skip_fields[field] = true
+end
+
+prefix = l.P(tag_prefix)
+chars = l.alnum + l.S("_-./")
+grammar = prefix * l.C(l.alpha * chars^0)
+
+buffer = {}
+last_flush = 0
+
+function flush()
+  add_to_payload(cjson.encode(buffer))
+  inject_payload()
+  last_flush = os.time()
+  buffer = {}
+end
+
+
+function process_message()
+
+  local ts
+  if ts_from_message then
+    ts = read_message("Timestamp") / 1e9
+  else
+    ts = os.time()
+  end
+
+  if not ts then return -1 end
+
+  local typ=read_message("Type")
+  local tags = {}
+  local metrics = {}
+
+  local msg = decode_message(read_message("raw"))
+  if not msg.Fields then return -1 end
+
+  for _, field in ipairs(msg.Fields) do
+    local name = field["name"]      -- these are related to way Heka works
+    local value = field["value"][1] --
+
+    local tag = grammar:match(name) or (tag_list[name] ~= nil and name)
+    if tag then
+      tags[tag] = value
+    else
+      if type(value) == "number" and skip_fields[name] == nil then
+        if type_as_prefix then
+          name = typ .. "." .. name
```

Comment:
Our policy is to only use the `..` concatenation operator for single concatentations. Any time you'd use more than one `..`, you should use `string.format` instead.